nx/lib/nx/binary_backend.ex

```diff
@@ -521,12 +521,36 @@ defmodule Nx.BinaryBackend do
   end
 
   defp bin_dot(%{type: t1} = left, contract_axes1, %{type: t2} = right, contract_axes2, type) do
-    bin_zip_reduce(left, contract_axes1, right, contract_axes2, type, 0, fn lhs, rhs, acc ->
+    {left, left_contract_axes} = bin_dot_transpose_contract_axes(left, contract_axes1)
+
+    {right, right_contract_axes} = bin_dot_transpose_contract_axes(right, contract_axes2)
+
+    bin_zip_reduce(left, left_contract_axes, right, right_contract_axes, type, 0, fn lhs,
+                                                                                     rhs,
+                                                                                     acc ->
       res = binary_to_number(lhs, t1) * binary_to_number(rhs, t2) + acc
       {res, res}
     end)
   end
 
+  defp bin_dot_transpose_contract_axes(tensor, contract_axes) do
+    remaining_axes =
+      contract_axes
+      |> Enum.sort(:desc)
+      |> Enum.reduce(Nx.axes(tensor), &List.delete_at(&2, &1))
+
+    transposed = Nx.transpose(tensor, axes: contract_axes ++ remaining_axes)
```

Comment:
@josevalim is it bad that I'm calling Nx.transpose and Nx.reshape here?
Should I call the BinaryBackend equivalents?

Reply:
That would be preferred, yes!